File: forge/src/main/java/generations/gg/generations/core/generationscore/forge/GenerationsForgeNetworkManager.java
package generations.gg.generations.core.generationscore.forge;

import generations.gg.generations.core.generationscore.GenerationsCore;
import generations.gg.generations.core.generationscore.GenerationsImplementation;
import generations.gg.generations.core.generationscore.network.ClientNetworkPacketHandler;
import generations.gg.generations.core.generationscore.network.GenerationsNetwork;
import generations.gg.generations.core.generationscore.network.ServerNetworkPacketHandler;
import generations.gg.generations.core.generationscore.network.packets.GenerationsNetworkPacket;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;
import net.minecraftforge.network.NetworkDirection;
import net.minecraftforge.network.NetworkRegistry;
import net.minecraftforge.network.PacketDistributor;
import net.minecraftforge.network.simple.SimpleChannel;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.function.BiConsumer;
import java.util.function.Function;

public class GenerationsForgeNetworkManager implements GenerationsImplementation.NetworkManager {
    public static final GenerationsForgeNetworkManager INSTANCE = new GenerationsForgeNetworkManager();

    private static final String PROTOCOL_VERSION = "1";
    private int id = 0;

    private SimpleChannel channel = NetworkRegistry.newSimpleChannel(
            GenerationsCore.id("main"),
            () -> PROTOCOL_VERSION,
            PROTOCOL_VERSION::equals,
            PROTOCOL_VERSION::equals
    );

    public void registerClientBound() {
        GenerationsNetwork.INSTANCE.registerClientBound();
    }

    public void registerServerBound() {
        GenerationsNetwork.INSTANCE.registerServerBound();
    }

    public <T extends GenerationsNetworkPacket<T>> void createClientBound(
            ResourceLocation identifier,
            Class<T> kClass,
            BiConsumer<T, FriendlyByteBuf> encoder,
            Function<FriendlyByteBuf, T> decoder,
            ClientNetworkPacketHandler<T> handler
    ) {
        this.channel.registerMessage(this.id++, kClass, encoder, decoder, (msg, ctx) -> {
            var context = ctx.get();
            handler.handleOnNettyThread(msg);
            context.setPacketHandled(true);
        });
    }

    public <T extends GenerationsNetworkPacket<T>> void createServerBound(
            ResourceLocation identifier,
            Class<T> kClass,
            BiConsumer<T, FriendlyByteBuf> encoder,
            Function<FriendlyByteBuf, T> decoder,
            ServerNetworkPacketHandler<T> handler
    ) {
        this.channel.registerMessage(this.id++, kClass, encoder, decoder, (msg, ctx) -> {
            var context = ctx.get();
            handler.handleOnNettyThread(msg, Objects.requireNonNull(Objects.requireNonNull(context.getSender()).getServer()), context.getSender());
            context.setPacketHandled(true);
        });
    }

    public void sendPacketToPlayer(@NotNull ServerPlayer player, @NotNull GenerationsNetworkPacket<?> packet) {
        this.channel.send(PacketDistributor.PLAYER.with(() -> player), packet);
    }

    public void sendPacketToServer(GenerationsNetworkPacket<?> packet) {
        this.channel.sendToServer(packet);
    }

    public <T extends GenerationsNetworkPacket<?>> @NotNull Packet<ClientGamePacketListener> asVanillaClientBound(T packet) {
        return (Packet<ClientGamePacketListener>) this.channel.toVanillaPacket(packet, NetworkDirection.PLAY_TO_CLIENT);
    }

    @Override
    public <T extends GenerationsNetworkPacket<?>, V extends Entity> void sendToAllTracking(T packet, V entity) {
        this.channel.send(PacketDistributor.TRACKING_ENTITY.with(() -> entity), packet);
    }
}

File: forge/src/main/java/generations/gg/generations/core/generationscore/forge/GenerationsForgeNetworkManager.kt
package generations.gg.generations.core.generationscore.forge

import dev.architectury.utils.Env
import dev.architectury.utils.EnvExecutor
import generations.gg.generations.core.generationscore.common.GenerationsCore
import generations.gg.generations.core.generationscore.common.GenerationsImplementation
import generations.gg.generations.core.generationscore.common.network.GenerationsNetwork
import generations.gg.generations.core.generationscore.common.network.ServerNetworkPacketHandler
import generations.gg.generations.core.generationscore.common.network.packets.GenerationsNetworkPacket
import net.minecraft.client.Minecraft
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.protocol.Packet
import net.minecraft.network.protocol.game.ClientGamePacketListener
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraftforge.fml.LogicalSide
import net.minecraftforge.network.NetworkDirection
import net.minecraftforge.network.NetworkEvent
import net.minecraftforge.network.NetworkRegistry
import net.minecraftforge.network.PacketDistributor
import java.util.*
import java.util.function.BiConsumer
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Supplier
import kotlin.reflect.KClass

object GenerationsForgeNetworkManager : GenerationsImplementation.NetworkManager {
    private var id = 0
    private val channel = NetworkRegistry.newSimpleChannel(
        GenerationsCore.id("main"),
        { PROTOCOL_VERSION },
        { anObject: String? -> PROTOCOL_VERSION.equals(anObject) }) { anObject: String? ->
        PROTOCOL_VERSION.equals(
            anObject
        )
    }

    override fun registerClientBound() {
        GenerationsNetwork.registerClientBound()
    }

    override fun registerServerBound() {
        GenerationsNetwork.registerServerBound()
    }

    override fun <T : GenerationsNetworkPacket<T>> createClientBound(
        identifier: ResourceLocation,
        kClass: KClass<T>,
        encoder: BiConsumer<T, FriendlyByteBuf>,
        decoder: Function<FriendlyByteBuf, T>,
        handler: Consumer<T>
    ) {
        channel.registerMessage(
            id++,
            kClass.java,
            encoder,
            decoder
        ) { msg: T, ctx: Supplier<NetworkEvent.Context> ->
            val context = ctx.get()
            handler.accept(msg)
            context.packetHandled = true
        }
    }



    override fun <T : GenerationsNetworkPacket<T>> createServerBound(
        identifier: ResourceLocation,
        kClass: KClass<T>,
        encoder: BiConsumer<T, FriendlyByteBuf>,
        decoder: Function<FriendlyByteBuf, T>,
        handler: ServerNetworkPacketHandler<T>
    ) {
        channel.registerMessage(id++, kClass.java, encoder, decoder) { msg: T, ctx: Supplier<NetworkEvent.Context> ->
            val context = ctx.get()
            handler.handleOnNettyThread(
                msg,
                Objects.requireNonNull(Objects.requireNonNull(context.sender)?.getServer()),
                context.sender
            )
            context.packetHandled = true
        }
    }

    override fun <T : GenerationsNetworkPacket<T>> createBothBound(
        identifier: ResourceLocation?,
        kClass: KClass<T>,
        encoder: BiConsumer<T, FriendlyByteBuf>,
        decoder: Function<FriendlyByteBuf, T>,
        clientHandler: Consumer<T>,
        serverHandler: ServerNetworkPacketHandler<T>,
    ) {
        channel.registerMessage(id++, kClass.java, encoder, decoder) { msg: T, ctx: Supplier<NetworkEvent.Context> ->
            val context = ctx.get()

            if(context.direction.receptionSide == LogicalSide.SERVER) {
                clientHandler.accept(msg)
            } else {
                serverHandler.handleOnNettyThread(
                    msg,
                    Objects.requireNonNull(Objects.requireNonNull(context.sender)?.getServer()),
                    context.sender
                )
            }
            context.packetHandled = true
        }
    }

    override fun sendPacketToPlayer(player: ServerPlayer, packet: GenerationsNetworkPacket<*>) {
        channel.send(PacketDistributor.PLAYER.with { player }, packet)
    }

    override fun sendPacketToServer(packet: GenerationsNetworkPacket<*>?) {
        channel.sendToServer(packet)
    }

    override fun <T : GenerationsNetworkPacket<*>?> asVanillaClientBound(packet: T): Packet<ClientGamePacketListener> {
        return channel.toVanillaPacket(packet, NetworkDirection.PLAY_TO_CLIENT) as Packet<ClientGamePacketListener>
    }

    override fun <T : GenerationsNetworkPacket<*>?, V : Entity?> sendToAllTracking(packet: T, entity: V) {
        channel.send(PacketDistributor.TRACKING_ENTITY.with { entity }, packet)
    }

    private const val PROTOCOL_VERSION = "1"
}